src/main/scala/cognite/spark/v1/FileContentRelation.scala

```diff
@@ -0,0 +1,108 @@
+import fs2.Stream
+import org.apache.commons.io.FileUtils
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.sources.{BaseRelation, Filter, PrunedFilteredScan, TableScan}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Row, SQLContext, SparkSession}
+import org.apache.spark.{Partition, TaskContext}
+import sttp.client3.{HttpURLConnectionBackend, UriContext, basicRequest}
+
+import java.net.URL
+import scala.io.Source
+
+trait WithSizeLimit {
+  val sizeLimit: Long
+}
+
+class FileContentRelation(config: RelationConfig, fileExternalId: String)(
+    override val sqlContext: SQLContext)
+    extends BaseRelation
+    with TableScan
+    with PrunedFilteredScan
+    with Serializable
+    with WithSizeLimit {
+
+  override val sizeLimit: Long = 5 * FileUtils.ONE_GB
+
+  @transient lazy val client: GenericClient[IO] =
+    CdpConnector.clientFromConfig(config)
+
+  override def schema: StructType =
+    createDataFrame(sqlContext.sparkSession).schema
+
+  def createDataFrame(sparkSession: SparkSession): DataFrame = {
+    val rowsRdd: RDD[String] = new RDD[String](sparkSession.sparkContext, Nil) with Serializable {
+
+      import cognite.spark.v1.CdpConnector.ioRuntime
+
+      val maxParallelism = 1
+      override def compute(split: Partition, context: TaskContext): Iterator[String] = {
+
+        val validUrl = for {
+          mimeType <- client.files.retrieveByExternalId(fileExternalId).map(_.mimeType)
+          downloadLink <- client.files
+            .downloadLink(FileDownloadExternalId(fileExternalId))
+            .map(_.downloadUrl)
+          isFileWithinLimits <- isFileWithinLimits(downloadLink)
+        } yield {
+          if (!isFileWithinLimits)
+            throw new CdfSparkException("File size above 5Gb limit")
+          if (mimeType.isEmpty || !mimeType.contains("application/jsonlines"))
+            throw new CdfSparkException("Wrong mimetype. Expects application/jsonlines")
+          else
+            downloadLink
+        }
+
+        StreamIterator(
+          Stream
+            .eval(validUrl)
+            .flatMap(readUrlContent),
+          maxParallelism,
+          None
+        )
+
+      }
+
+      override protected def getPartitions: Array[Partition] =
+        Array(CdfPartition(0))
+    }
+    import sparkSession.implicits._
+    sparkSession.read.json(rowsRdd.toDS())
+  }
+
+  private def readUrlContent(url: String): Stream[IO, String] =
+    Stream
+      .bracket(IO(new URL(url).openStream())) { inStream =>
+        IO(inStream.close()) // Ensure the InputStream is closed properly
+      }
+      .flatMap { inStream =>
+        val source = Source.fromInputStream(inStream)
+        Stream.fromIterator[IO](source.getLines(), chunkSize = 4096)
+      }
+
+  private def isFileWithinLimits(downloadUrl: String): IO[Boolean] = {
+    val backend = Resource.make(IO(HttpURLConnectionBackend()))(b => IO(b.close()))
```

Comment:
HttpURLConnectionBackend is odd and doesn't quite fit here..
- it's synchronous so should be wrapped in IO.blocking if put into IO
- we do have shared sttpBackend most other places, though it could be hidden within sdk client, but some if it's construction isn't hidden in CdpConnection, so might borrow some stuff (if not switch entirely to sdk client)
- sdk client would also have some features that it lacks: tracing (we'll see only request going out to blob storage I guess), maybe some metrics, retry policies and rate limiters etc

Reply:
Hmm, I agree and for the head request it's fairly straightforwards to just add a head method which returns the headers to the SDK. For the get method it's a bit more involved as we are using .openStream() but maybe we can do that more explicitely, I'll investigate